### ISSUE TITLE
Phase 33.3 — ReplayMemoryManager: Experience replay with prioritized sampling and generative rehearsal

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#693]
## Overview

The `ReplayMemoryManager` implements experience replay strategies for continual learning, storing and replaying past experiences to prevent catastrophic forgetting. It supports multiple replay paradigms: raw experience replay with prioritized sampling, generative replay via pseudo-rehearsal (Shin et al., 2017), coreset selection for memory-efficient storage, and dark experience replay storing logits for knowledge distillation (Buzzega et al., 2020).

## Key Responsibilities

- **Priority replay buffer** — Experience storage with TD-error or loss-based priority sampling (Schaul et al., 2015 adapted for CL)
- **Generative replay** — Train generative model (VAE/GAN) to produce pseudo-samples of previous tasks
- **Coreset selection** — Select maximally representative subsets via k-center, herding, or gradient matching
- **Dark experience replay** — Store model logits alongside inputs for knowledge-distillation-based replay
- **Memory budget allocation** — Distribute limited buffer capacity across tasks using configurable strategies
- **Replay scheduling** — Control replay frequency, batch composition (ratio of new vs. replayed data)
- **Forgetting-aware sampling** — Increase replay probability for tasks showing accuracy degradation

### Inputs
- `buffer_size: int` — Maximum number of stored experiences (default: 5000)
- `replay_strategy: Literal["random", "prioritized", "generative", "coreset", "dark"]`
- `task_ratio: float` — Fraction of each training batch drawn from replay buffer (default: 0.5)
- `generative_model: Optional[nn.Module]` — Generator for pseudo-rehearsal
- `coreset_method: Literal["k_center", "herding", "gradient_matching"]`
- `dark_alpha: float` — Distillation temperature for dark replay (default: 0.5)

### Outputs
- `ReplayBatch` — Mixed batch of new + replayed experiences with metadata
- `BufferStats` — Per-task sample counts, priority distributions, memory usage
- `ReplayDiagnostics` — Replay effectiveness metrics, forgetting correlation analysis

## Acceptance Criteria

- [ ] Prioritized replay reduces forgetting by >20% vs. random replay (Split-CIFAR10)
- [ ] Generative replay matches raw replay performance with <10% stored samples
- [ ] Coreset selection retains 90% of full-buffer performance at 10% buffer size
- [ ] Dark experience replay outperforms naive replay on class-incremental scenarios
- [ ] Buffer operations (add/sample) O(log n) with priority queue
- [ ] Memory budget strictly enforced with graceful eviction
- [ ] Thread-safe for concurrent training and replay sampling
- [ ] Integration tests with Phase 33.5 ContinualOrchestrator
- [ ] 95%+ test coverage, all public methods typed

## References

- Shin et al. (2017) — Continual Learning with Deep Generative Replay
- Buzzega et al. (2020) — Dark Experience Replay
- Lopez-Paz & Ranzato (2017) — Gradient Episodic Memory (GEM)
- Chaudhry et al. (2019) — Efficient Lifelong Learning with A-GEM
- Rebuffi et al. (2017) — iCaRL: Incremental Classifier and Representation Learning
- Planning: #690

**Comment by web3guru888** [comment]
## Implementation Notes — ReplayMemoryManager

### Python Skeleton

```python
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Any, Dict, List, Literal, Optional, Tuple
import heapq, random, threading
import torch
from torch import Tensor, nn

@dataclass
class Experience:
    """Single stored experience."""
    input: Tensor
    target: Tensor
    task_id: str
    logits: Optional[Tensor] = None   # for dark experience replay
    priority: float = 1.0
    age: int = 0                       # number of tasks since storage

@dataclass
class ReplayBatch:
    inputs: Tensor
    targets: Tensor
    task_ids: List[str]
    logits: Optional[Tensor]
    is_replay: Tensor                  # bool mask: True = replayed

@dataclass
class BufferStats:
    total_stored: int
    per_task_counts: Dict[str, int]
    memory_bytes: int
    priority_mean: float
    priority_std: float

class ReplayMemoryManager:
    """Experience replay for continual learning."""

    def __init__(
        self,
        buffer_size: int = 5000,
        replay_strategy: Literal["random", "prioritized", "generative", "coreset", "dark"] = "prioritized",
        task_ratio: float = 0.5,
        generative_model: Optional[nn.Module] = None,
        coreset_method: Literal["k_center", "herding", "gradient_matching"] = "herding",
        dark_alpha: float = 0.5,
    ) -> None:
        self._max_size = buffer_size
        self._strategy = replay_strategy
        self._task_ratio = task_ratio
        self._generator = generative_model
        self._coreset_method = coreset_method
        self._dark_alpha = dark_alpha
        self._buffer: List[Experience] = []
        self._lock = threading.Lock()

    # ── public API ──────────────────────────────────────────────
    def add(self, experience: Experience) -> None:
        """Add experience to buffer, evicting lowest-priority if full."""
        with self._lock:
            if len(self._buffer) >= self._max_size:
                self._evict()
            self._buffer.append(experience)

    def add_batch(self, inputs: Tensor, targets: Tensor, task_id: str,
                  logits: Optional[Tensor] = None) -> int:
        """Convenience: store a batch of experiences."""
        added = 0
        for i in range(inputs.size(0)):
            exp = Experience(
                input=inputs[i], target=targets[i], task_id=task_id,
                logits=logits[i] if logits is not None else None,
            )
            self.add(exp)
            added += 1
        return added

    def sample(self, batch_size: int) -> ReplayBatch:
        """Sample a replay batch according to the active strategy."""
        with self._lock:
            if self._strategy == "prioritized":
                return self._prioritized_sample(batch_size)
            elif self._strategy == "generative":
                return self._generative_sample(batch_size)
            elif self._strategy == "dark":
                return self._dark_sample(batch_size)
            elif self._strategy == "coreset":
                return self._coreset_sample(batch_size)
            else:
                return self._random_sample(batch_size)

    def mixed_batch(self, new_inputs: Tensor, new_targets: Tensor,
                    task_id: str) -> ReplayBatch:
        """Create a mixed batch: (1-ratio) new + ratio replayed."""
        total = new_inputs.size(0)
        n_replay = int(total * self._task_ratio)
        n_new = total - n_replay
        replay = self.sample(n_replay)
        inputs = torch.cat([new_inputs[:n_new], replay.inputs])
        targets = torch.cat([new_targets[:n_new], replay.targets])
        is_replay = torch.cat([
            torch.zeros(n_new, dtype=torch.bool),
            torch.ones(n_replay, dtype=torch.bool),
        ])
        return ReplayBatch(inputs=inputs, targets=targets,
                           task_ids=[task_id]*n_new + replay.task_ids,
                           logits=replay.logits, is_replay=is_replay)

    def update_priorities(self, indices: List[int], new_priorities: List[float]) -> None:
        """Update priorities for experiences (e.g., after loss computation)."""
        with self._lock:
            for idx, p in zip(indices, new_priorities):
                if 0 <= idx < len(self._buffer):
                    self._buffer[idx].priority = p

    def stats(self) -> BufferStats:
        """Return buffer statistics."""
        counts: Dict[str, int] = {}
        priorities = []
        for exp in self._buffer:
            counts[exp.task_id] = counts.get(exp.task_id, 0) + 1
            priorities.append(exp.priority)
        p_tensor = torch.tensor(priorities) if priorities else torch.zeros(1)
        return BufferStats(
            total_stored=len(self._buffer),
            per_task_counts=counts,
            memory_bytes=sum(e.input.nelement() * e.input.element_size() for e in self._buffer),
            priority_mean=p_tensor.mean().item(),
            priority_std=p_tensor.std().item(),
        )

    # ── selection strategies ────────────────────────────────────
    def select_coreset(self, task_data: List[Experience], k: int) -> List[Experience]:
        """Select k representative samples using configured coreset method."""
        if self._coreset_method == "herding":
            return self._herding_select(task_data, k)
        elif self._coreset_method == "k_center":
            return self._k_center_select(task_data, k)
        else:
            return self._gradient_matching_select(task_data, k)

    # ── internals ───────────────────────────────────────────────
    def _evict(self) -> None:
        min_idx = min(range(len(self._buffer)), key=lambda i: self._buffer[i].priority)
        self._buffer.pop(min_idx)

    def _prioritized_sample(self, n: int) -> ReplayBatch: ...
    def _generative_sample(self, n: int) -> ReplayBatch: ...
    def _dark_sample(self, n: int) -> ReplayBatch: ...
    def _coreset_sample(self, n: int) -> ReplayBatch: ...
    def _random_sample(self, n: int) -> ReplayBatch: ...
    def _herding_select(self, data: List[Experience], k: int) -> List[Experience]: ...
    def _k_center_select(self, data: List[Experience], k: int) -> List[Experience]: ...
    def _gradient_matching_select(self, data: List[Experience], k: int) -> List[Experience]: ...
```

### Algorithm — Dark Experience Replay (Buzzega et al. 2020)

```
procedure DARK_EXPERIENCE_REPLAY(model, new_batch, buffer, α):
    // Store current logits for new data
    with no_grad:
        new_logits ← model(new_batch.inputs)
    buffer.add(new_batch.inputs, new_batch.targets, logits=new_logits)

    // Sample from buffer
    replay ← buffer.sample(|new_batch|)

    // Standard loss on new data
    L_new ← cross_entropy(model(new_batch.inputs), new_batch.targets)

    // Knowledge distillation loss on replayed data
    replay_out ← model(replay.inputs)
    L_dark ← α · MSE(replay_out, replay.logits) + (1-α) · cross_entropy(replay_out, replay.targets)

    return L_new + L_dark
```

### Type Table (mypy-compatible)

| Symbol | Type | Description |
|--------|------|-------------|
| `buffer_size` | `int` | Maximum buffer capacity |
| `replay_strategy` | `Literal[...]` | Sampling strategy |
| `task_ratio` | `float` | Fraction of batch from replay |
| `Experience.logits` | `Optional[Tensor]` | Stored logits for dark replay |
| `Experience.priority` | `float` | Sampling priority |
| `ReplayBatch.is_replay` | `Tensor` | Boolean mask for replayed samples |
| `BufferStats.per_task_counts` | `Dict[str, int]` | Samples per task |
| `dark_alpha` | `float` | Distillation weight [0, 1] |
